src/util/amount.rs

```diff
@@ -0,0 +1,789 @@
+        }
+    }
+}
+
+// The inner type used to represent amounts.
+// This is done to more easily change the underlying type in case this is desired in the future.
+type Inner = i64;
+
+/// Type to represent Bitcoin amounts.
+#[derive(Copy, Clone, Hash)]
+pub struct Amount(Inner);
+// The Inner amount represents the number of satoshis.
+
+impl Amount {
+    /// Create a new Amount using `amount` as the Inner type.
+    fn from_inner(amount: Inner) -> Amount {
+        Amount(amount)
+    }
+
+    /// The zero amount.
+    pub fn zero() -> Amount {
+        Amount::from_inner(0)
+    }
+
+    /// Create an Amount with satoshi precision and the given number of satoshis.
+    pub fn from_sat(satoshi: i64) -> Amount {
+        Amount::from_inner(satoshi)
+    }
+
+    /// Get the number of satoshis in this amount.
+    pub fn as_sat(self) -> i64 {
+        self.0
+    }
+
+    /// The maximum value of an Amount.
+    pub fn max_value() -> Amount {
+        Amount::from_inner(Inner::max_value())
+    }
+
+    /// The minimum value of an amount.
+    pub fn min_value() -> Amount {
+        Amount::from_inner(Inner::min_value())
+    }
+
+    // Don't use the Inner type in the methods below.
+    // Always use [Amount::from_sat] and [Amount::as_sat] instead.
+
+    /// Convert from a value expressing bitcoins to an Amount.
+    pub fn from_btc<T: IntoBtc>(btc: T) -> Amount {
+        btc.into_btc()
+    }
+
+    /// Parse a decimal string as a value in the given denomination.
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value with denomination,
+    /// use `FromStr`.
+    pub fn parse_denom(mut s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
+        if s.len() == 0 {
+            return Err(ParseAmountError::InvalidFormat);
+        }
+
+        let negative = s.chars().nth(0).unwrap() == '-';
+        if negative {
+            if s.len() == 1 {
+                return Err(ParseAmountError::InvalidFormat);
+            }
+            s = &s[1..];
+        }
+
+        let max_decimals = {
+            // The difference in precision between native (satoshi)
+            // and desired denomination.
+            let precision_diff = -denom.precision();
+            if precision_diff < 0 {
+                // If precision diff is negative, this means we are parsing into a less
+                // precise amount.  That is not allowed unless the last digits are zeroes
+                // as many as the diffence in precision.
+                let last_n = precision_diff.abs() as usize;
+                if !s.chars().skip(s.len() - last_n).all(|d| d == '0') {
+                    return Err(ParseAmountError::TooPrecise);
+                }
+                s = &s[0..s.len() - last_n];
+                0
+            } else {
+                precision_diff
+            }
+        };
+
+        let mut decimals = None;
+        let mut value: i64 = 0; // as satoshis
+        for c in s.as_bytes() {
```

Comment:
Is there a good reason to use bytes instead of chars?

Reply:
Good question. Probably not. What's the exact difference? I think chars is probably safer, even though `String`s are guaranteed to be UTF8, right?

Reply:
using chars now